File: src/main/services/mod-manager/presets.ts
import path from "node:path";
import type { ApplyPresetResult, Preset } from "@shared/types";
import { trim } from "es-toolkit";
import { nanoid } from "nanoid";
import type { NahidaDesktop } from "../..";
import type { ModPresetItemRow, ModPresetRow } from "../../internal/db/schema";
import type { ModLibraryService } from "./library";
import type { ModActionsService } from "./mod-actions";
import {
    DISABLED_PREFIX_REGEX,
    normalizeModPath,
    renameWithUniqueName,
    restoreDisabledPrefix,
    stripDisabledPrefix,
    toGameRelativePath,
} from "./path-utils";

interface PresetSnapshotItemRecord {
    modKey: string;
    relativePath: string;
    groupRelativePath: string;
    folderName: string;
    isEnabled: boolean;
}

interface ScannedPresetItem extends PresetSnapshotItemRecord {
    actualPath: string;
}

interface PresetConflictCandidate {
    actualPath: string;
    relativePath: string;
    folderName: string;
    isEnabled: boolean;
}

export interface PresetConflict {
    modKey: string;
    candidates: PresetConflictCandidate[];
}

const MOD_PRESET_ITEM_INSERT_BATCH_SIZE = 100;
const MOD_PRESET_VERSION = 2;

export class ModPresetsService {
    constructor(
        private readonly desktop: NahidaDesktop,
        private readonly library: ModLibraryService,
        private readonly actions: ModActionsService,
    ) {}

    public async presets(game: string): Promise<Preset[]> {
        const results = await this.desktop.lib.db.modPresets.listByGame(game);

        return results
            .sort((a, b) => a.name.localeCompare(b.name))
            .map((r) => ({
                id: r.id,
                game: r.game,
                name: r.name,
                description: r.description ?? null,
                createdAt: r.createdAt,
                updatedAt: r.updatedAt,
                version: r.version,
                isLegacy: r.version < MOD_PRESET_VERSION,
            }));
    }

    public async presetCreateConflicts(game: string): Promise<PresetConflict[]> {
        return await this.getPresetSnapshotConflicts(game);
    }

    public async createPreset(
        game: string,
        name: string,
        description?: string,
        resolveConflicts = false,
    ): Promise<Preset> {
        const trimmedName = trim(name);
        const trimmedDescription = trim(description ?? "");
        if (!trimmedName) {
            throw new Error("INVALID_PRESET_NAME");
        }

        const existingPreset = await this.desktop.lib.db.modPresets.findByGameAndName(
            game,
            trimmedName,
        );
        if (existingPreset) {
            throw new Error("PRESET_NAME_EXISTS");
        }

        if (resolveConflicts) {
            const remainingConflicts = await this.resolvePresetSnapshotConflicts(game);
            if (remainingConflicts.length > 0) {
                throw new Error("PRESET_CONFLICT_RESOLUTION_FAILED");
            }
        } else {
            const conflicts = await this.getPresetSnapshotConflicts(game);
            if (conflicts.length > 0) {
                throw new Error("PRESET_CONFLICTS_EXIST");
            }
        }

        const snapshot = await this.getPresetSnapshot(game);

        const id = nanoid();
        const now = new Date().toISOString();

        this.desktop.lib.db.transaction(() => {
            this.desktop.lib.db.modPresets.insert({
                id,
                game,
                name: trimmedName,
                description: trimmedDescription || null,
                itemCount: 0,
                createdAt: now,
                updatedAt: now,
                version: MOD_PRESET_VERSION,
            } satisfies ModPresetRow);

            if (snapshot.length > 0) {
                const presetItems = snapshot.map((item, index) => ({
                    presetId: id,
                    ...this.toPresetRecord(item),
                    itemOrder: index,
                }));

                for (
                    let startIndex = 0;
                    startIndex < presetItems.length;
                    startIndex += MOD_PRESET_ITEM_INSERT_BATCH_SIZE
                ) {
                    this.desktop.lib.db.modPresetItems.insertMany(
                        presetItems.slice(
                            startIndex,
                            startIndex + MOD_PRESET_ITEM_INSERT_BATCH_SIZE,
                        ) satisfies ModPresetItemRow[],
                    );
                }
            }
        });

        return {
            id,
            game,
            name: trimmedName,
            description: trimmedDescription || null,
            createdAt: now,
            updatedAt: now,
            version: MOD_PRESET_VERSION,
            isLegacy: false,
        };
    }

    public async applyPreset(presetId: string): Promise<ApplyPresetResult> {
        const preset = await this.desktop.lib.db.modPresets.findById(presetId);

        if (!preset) {
            throw new Error(`Preset ${presetId} not found`);
        }

        if (preset.version < MOD_PRESET_VERSION) {
            throw new Error("LEGACY_PRESET_NOT_SUPPORTED");
        }

        const presetItems = await this.desktop.lib.db.modPresetItems.listByPresetId(presetId);
        const currentItems = await this.getPresetSnapshot(preset.game);
        const currentByKey = new Map(currentItems.map((item) => [item.modKey, item] as const));
        const currentByRelativePath = new Map(
            currentItems.map((item) => [item.relativePath.toLowerCase(), item] as const),
        );
        const result: ApplyPresetResult = {
            presetId,
            applied: [],
            skipped: [],
            missing: [],
        };

        for (const presetItem of presetItems.sort((a, b) => a.itemOrder - b.itemOrder)) {
            const currentItem =
                currentByKey.get(presetItem.modKey) ??
                currentByRelativePath.get(presetItem.relativePath.toLowerCase());

            if (!currentItem) {
                result.missing.push({
                    modKey: presetItem.modKey,
                    expectedFolderName: presetItem.folderName,
                    expectedRelativePath: presetItem.relativePath,
                });
                continue;
            }

            if (currentItem.isEnabled === presetItem.isEnabled) {
                result.skipped.push(currentItem.relativePath);
                continue;
            }

            try {
                if (presetItem.isEnabled) {
                    await this.actions.enable(currentItem.actualPath);
                } else {
                    await this.actions.disable(currentItem.actualPath);
                }
                result.applied.push(currentItem.relativePath);
            } catch (error) {
                this.desktop.logger.error(
                    error,
                    `Mod:applyPreset:${presetItem.isEnabled ? "enable" : "disable"}:${currentItem.actualPath}`,
                );
            }
        }

        return result;
    }

    public async deletePreset(presetId: string): Promise<void> {
        await this.desktop.lib.db.modPresets.delete(presetId);
    }

    public async updatePresetName(presetId: string, newName: string): Promise<void> {
        const preset = await this.desktop.lib.db.modPresets.findById(presetId);

        if (!preset) {
            throw new Error(`Preset ${presetId} not found`);
        }

        const trimmedName = trim(newName);
        if (!trimmedName) {
            throw new Error("INVALID_PRESET_NAME");
        }

        const existingPreset = await this.desktop.lib.db.modPresets.findByGameAndName(
            preset.game,
            trimmedName,
        );
        if (existingPreset && existingPreset.id !== presetId) {
            throw new Error("PRESET_NAME_EXISTS");
        }

        await this.desktop.lib.db.modPresets.updateName(
            presetId,
            trimmedName,
            new Date().toISOString(),
        );
    }

    private buildModKey(gamePath: string, groupPath: string, modPath: string): string {
        const groupRelativePath = toGameRelativePath(gamePath, groupPath);
        const modRelativePath = toGameRelativePath(gamePath, modPath);
        return `${groupRelativePath}::${modRelativePath}`;
    }

    private buildPresetSnapshotItem(
        gamePath: string,
        groupPath: string,
        modPath: string,
    ): ScannedPresetItem {
        const folderName = path.basename(modPath);
        return {
            modKey: this.buildModKey(gamePath, groupPath, modPath),
            relativePath: toGameRelativePath(gamePath, modPath),
            groupRelativePath: toGameRelativePath(gamePath, groupPath),
            folderName: stripDisabledPrefix(folderName),
            isEnabled: !DISABLED_PREFIX_REGEX.test(folderName),
            actualPath: modPath,
        };
    }

    private getPresetGroupPath(gamePath: string, modPath: string): string {
        const relativePath = path.relative(gamePath, modPath);
        const segments = relativePath.split(/[\\/]+/).filter(Boolean);

        if (segments.length <= 1) {
            return gamePath;
        }

        return path.join(gamePath, ...segments.slice(0, -1));
    }

    private async collectPresetSnapshotItems(gamePath: string): Promise<ScannedPresetItem[]> {
        const iniPaths = await this.desktop.lib.fs.findFiles(gamePath, {
            extensions: [".ini"],
        });
        const modPathMap = new Map<string, string>();

        for (const iniPath of iniPaths) {
            const modPath = path.dirname(iniPath);
            const normalizedModPath = normalizeModPath(modPath);
            if (!modPathMap.has(normalizedModPath)) {
                modPathMap.set(normalizedModPath, modPath);
            }
        }

        return Array.from(modPathMap.values())
            .sort((a, b) => a.localeCompare(b))
            .map((modPath) => {
                const groupPath = this.getPresetGroupPath(gamePath, modPath);
                return this.buildPresetSnapshotItem(gamePath, groupPath, modPath);
            });
    }

    private async getPresetSnapshot(game: string): Promise<ScannedPresetItem[]> {
        const gamePath = await this.library.gamePath(game);
        if (!gamePath) {
            throw new Error(`No mod folder path set for ${game}`);
        }

        const items = await this.collectPresetSnapshotItems(gamePath);
        return items
            .sort((a, b) => a.relativePath.localeCompare(b.relativePath))
            .map((item) => ({
                ...item,
                relativePath: item.relativePath,
            }));
    }

    private async getPresetSnapshotConflicts(game: string): Promise<PresetConflict[]> {
        const items = await this.getPresetSnapshot(game);
        const itemsByModKey = new Map<string, ScannedPresetItem[]>();

        for (const item of items) {
            const conflicts = itemsByModKey.get(item.modKey) ?? [];
            conflicts.push(item);
            itemsByModKey.set(item.modKey, conflicts);
        }

        return Array.from(itemsByModKey.entries())
            .filter(([, conflicts]) => conflicts.length > 1)
            .map(([modKey, conflicts]) => ({
                modKey,
                candidates: conflicts
                    .sort((a, b) => a.relativePath.localeCompare(b.relativePath))
                    .map((item) => ({
                        actualPath: item.actualPath,
                        relativePath: item.relativePath,
                        folderName: item.folderName,
                        isEnabled: item.isEnabled,
                    })),
            }))
            .sort((a, b) => a.modKey.localeCompare(b.modKey));
    }

    private async resolvePresetSnapshotConflicts(game: string): Promise<PresetConflict[]> {
        const conflicts = await this.getPresetSnapshotConflicts(game);

        for (const conflict of conflicts) {
            const candidates = conflict.candidates
                .map((candidate) =>
                    candidate.actualPath
                        ? {
                              ...candidate,
                              actualPath: candidate.actualPath,
                          }
                        : null,
                )
                .filter((candidate): candidate is PresetConflictCandidate => candidate !== null);
            const enabledCandidates = candidates.filter((candidate) => candidate.isEnabled);
            const disabledCandidates = candidates.filter((candidate) => !candidate.isEnabled);

            const renameTargets =
                enabledCandidates.length > 0
                    ? disabledCandidates
                    : disabledCandidates.length > 1
                      ? disabledCandidates.slice(1)
                      : [];

            for (const candidate of renameTargets) {
                await renameWithUniqueName(
                    this.desktop.lib.fs,
                    candidate.actualPath,
                    restoreDisabledPrefix(
                        path.basename(candidate.actualPath),
                        candidate.folderName,
                    ),
                );
            }
        }

        return await this.getPresetSnapshotConflicts(game);
    }

    private toPresetRecord(item: ScannedPresetItem): PresetSnapshotItemRecord {
        return {
            modKey: item.modKey,
            relativePath: item.relativePath,
            groupRelativePath: item.groupRelativePath,
            folderName: item.folderName,
            isEnabled: item.isEnabled,
        };
    }
}

File: internal/mod/presets.go
package mod

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"io/fs"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"nahida.live/desktop/internal/db"
)

const modPresetVersion int64 = 2

type Preset struct {
	ID          string  `json:"id"`
	Game        string  `json:"game"`
	Name        string  `json:"name"`
	Description *string `json:"description"`
	CreatedAt   string  `json:"createdAt"`
	UpdatedAt   string  `json:"updatedAt"`
	Version     int64   `json:"version"`
	IsLegacy    bool    `json:"isLegacy"`
}

type PresetConflictCandidate struct {
	ActualPath   string `json:"actualPath"`
	RelativePath string `json:"relativePath"`
	FolderName   string `json:"folderName"`
	IsEnabled    bool   `json:"isEnabled"`
}

type PresetConflict struct {
	ModKey     string                    `json:"modKey"`
	Candidates []PresetConflictCandidate `json:"candidates"`
}

type MissingPresetItem struct {
	ModKey               string `json:"modKey"`
	ExpectedFolderName   string `json:"expectedFolderName"`
	ExpectedRelativePath string `json:"expectedRelativePath"`
}

type ApplyPresetResult struct {
	PresetID string              `json:"presetId"`
	Applied  []string            `json:"applied"`
	Skipped  []string            `json:"skipped"`
	Missing  []MissingPresetItem `json:"missing"`
}

type presetSnapshotItem struct {
	ModKey            string
	RelativePath      string
	GroupRelativePath string
	FolderName        string
	IsEnabled         bool
	ActualPath        string
}

func (m *Mod) GetPresets(ctx context.Context, game string) ([]Preset, error) {
	client, err := m.requireClient()
	if err != nil {
		return nil, err
	}
	rows, err := client.ModPresets.ListByGame(ctx, game)
	if err != nil {
		return nil, err
	}
	result := make([]Preset, len(rows))
	for i := range rows {
		result[i] = presetFromRow(rows[i])
	}
	less := newLocaleLess()
	sort.SliceStable(result, func(i, j int) bool { return less(result[i].Name, result[j].Name) })
	return result, nil
}

func (m *Mod) GetPresetCreateConflicts(ctx context.Context, game string) ([]PresetConflict, error) {
	items, err := m.presetSnapshot(ctx, game)
	if err != nil {
		return nil, err
	}
	return snapshotConflicts(items), nil
}

func (m *Mod) CreatePreset(
	ctx context.Context,
	game, name string,
	description *string,
	resolveConflicts bool,
) (Preset, error) {
	client, err := m.requireClient()
	if err != nil {
		return Preset{}, err
	}
	name = strings.TrimSpace(name)
	if name == "" {
		return Preset{}, errors.New("INVALID_PRESET_NAME")
	}
	description = cleanOptional(description)
	existing, err := client.ModPresets.FindByGameAndName(ctx, game, name)
	if err != nil {
		return Preset{}, err
	}
	if existing != nil {
		return Preset{}, errors.New("PRESET_NAME_EXISTS")
	}
	items, err := m.presetSnapshot(ctx, game)
	if err != nil {
		return Preset{}, err
	}
	conflicts := snapshotConflicts(items)
	if len(conflicts) > 0 && !resolveConflicts {
		return Preset{}, errors.New("PRESET_CONFLICTS_EXIST")
	}
	if len(conflicts) > 0 {
		if err := resolveSnapshotConflicts(conflicts); err != nil {
			return Preset{}, err
		}
		items, err = m.presetSnapshot(ctx, game)
		if err != nil {
			return Preset{}, err
		}
		if len(snapshotConflicts(items)) > 0 {
			return Preset{}, errors.New("PRESET_CONFLICT_RESOLUTION_FAILED")
		}
	}
	id, err := newPresetID()
	if err != nil {
		return Preset{}, err
	}
	now := time.Now().UTC().Format(time.RFC3339Nano)
	row := db.ModPresetRow{
		ID: id, Game: game, Name: name, Description: description, ItemCount: int64(len(items)),
		CreatedAt: now, UpdatedAt: now, Version: modPresetVersion,
	}
	dbItems := make([]db.ModPresetItemRow, len(items))
	for i, item := range items {
		dbItems[i] = db.ModPresetItemRow{
			PresetID: id, ModKey: item.ModKey, RelativePath: item.RelativePath,
			GroupRelativePath: item.GroupRelativePath, FolderName: item.FolderName,
			IsEnabled: item.IsEnabled, ItemOrder: int64(i),
		}
	}
	if err := client.ModPresets.InsertSnapshot(ctx, row, dbItems); err != nil {
		return Preset{}, err
	}
	return presetFromRow(row), nil
}

func (m *Mod) ApplyPreset(ctx context.Context, presetID string) (ApplyPresetResult, error) {
	client, err := m.requireClient()
	if err != nil {
		return ApplyPresetResult{}, err
	}
	preset, err := client.ModPresets.FindByID(ctx, presetID)
	if err != nil {
		return ApplyPresetResult{}, err
	}
	if preset == nil {
		return ApplyPresetResult{}, fmt.Errorf("preset %s not found", presetID)
	}
	if preset.Version < modPresetVersion {
		return ApplyPresetResult{}, errors.New("LEGACY_PRESET_NOT_SUPPORTED")
	}
	stored, err := client.ModPresetItems.ListByPresetID(ctx, presetID)
	if err != nil {
		return ApplyPresetResult{}, err
	}
	current, err := m.presetSnapshot(ctx, preset.Game)
	if err != nil {
		return ApplyPresetResult{}, err
	}
	byKey := make(map[string]presetSnapshotItem, len(current))
	byRelative := make(map[string]presetSnapshotItem, len(current))
	for _, item := range current {
		byKey[item.ModKey] = item
		byRelative[strings.ToLower(item.RelativePath)] = item
	}
	result := ApplyPresetResult{
		PresetID: presetID, Applied: []string{}, Skipped: []string{}, Missing: []MissingPresetItem{},
	}
	sort.Slice(stored, func(i, j int) bool { return stored[i].ItemOrder < stored[j].ItemOrder })
	for _, wanted := range stored {
		item, ok := byKey[wanted.ModKey]
		if !ok {
			item, ok = byRelative[strings.ToLower(wanted.RelativePath)]
		}
		if !ok {
			result.Missing = append(result.Missing, MissingPresetItem{
				ModKey: wanted.ModKey, ExpectedFolderName: wanted.FolderName,
				ExpectedRelativePath: wanted.RelativePath,
			})
			continue
		}
		if item.IsEnabled == wanted.IsEnabled {
			result.Skipped = append(result.Skipped, item.RelativePath)
			continue
		}
		if m.isActiveDownloadDestination(item.ActualPath) {
			result.Skipped = append(result.Skipped, item.RelativePath)
			continue
		}
		var actionErr error
		if wanted.IsEnabled {
			_, actionErr = m.Enable(ctx, item.ActualPath)
		} else {
			_, actionErr = m.Disable(ctx, item.ActualPath)
		}
		if actionErr != nil {
			if actionErr.Error() == "MOD_DOWNLOAD_IN_PROGRESS" {
				result.Skipped = append(result.Skipped, item.RelativePath)
			}
			continue
		}
		result.Applied = append(result.Applied, item.RelativePath)
	}
	return result, nil
}

func (m *Mod) DeletePreset(ctx context.Context, presetID string) error {
	client, err := m.requireClient()
	if err != nil {
		return err
	}
	return client.ModPresets.Delete(ctx, presetID)
}

func (m *Mod) UpdatePresetName(ctx context.Context, presetID, newName string) error {
	client, err := m.requireClient()
	if err != nil {
		return err
	}
	preset, err := client.ModPresets.FindByID(ctx, presetID)
	if err != nil {
		return err
	}
	if preset == nil {
		return fmt.Errorf("preset %s not found", presetID)
	}
	newName = strings.TrimSpace(newName)
	if newName == "" {
		return errors.New("INVALID_PRESET_NAME")
	}
	existing, err := client.ModPresets.FindByGameAndName(ctx, preset.Game, newName)
	if err != nil {
		return err
	}
	if existing != nil && existing.ID != presetID {
		return errors.New("PRESET_NAME_EXISTS")
	}
	return client.ModPresets.UpdateName(ctx, presetID, newName, time.Now().UTC().Format(time.RFC3339Nano))
}

func (m *Mod) presetSnapshot(ctx context.Context, game string) ([]presetSnapshotItem, error) {
	gamePath, err := m.GetGamePath(ctx, game)
	if err != nil {
		return nil, err
	}
	if gamePath == nil {
		return nil, fmt.Errorf("no mod folder path set for %s", game)
	}
	paths := map[string]string{}
	err = filepath.WalkDir(*gamePath, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if entry.Type().IsRegular() && strings.EqualFold(filepath.Ext(path), ".ini") {
			parent := filepath.Dir(path)
			key := strings.ToLower(filepath.Clean(parent))
			if _, exists := paths[key]; !exists {
				paths[key] = parent
			}
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	items := make([]presetSnapshotItem, 0, len(paths))
	for _, modPath := range paths {
		groupPath := filepath.Dir(modPath)
		groupRelative := normalizedGameRelativePath(*gamePath, groupPath)
		relative := normalizedGameRelativePath(*gamePath, modPath)
		items = append(items, presetSnapshotItem{
			ModKey: groupRelative + "::" + relative, RelativePath: relative,
			GroupRelativePath: groupRelative, FolderName: stripDisabled(filepath.Base(modPath)),
			IsEnabled: !isDisabled(filepath.Base(modPath)), ActualPath: modPath,
		})
	}
	less := newLocaleLess()
	sort.SliceStable(items, func(i, j int) bool { return less(items[i].RelativePath, items[j].RelativePath) })
	return items, nil
}

func snapshotConflicts(items []presetSnapshotItem) []PresetConflict {
	byKey := map[string][]presetSnapshotItem{}
	for _, item := range items {
		byKey[item.ModKey] = append(byKey[item.ModKey], item)
	}
	result := []PresetConflict{}
	for key, values := range byKey {
		if len(values) < 2 {
			continue
		}
		less := newLocaleLess()
		sort.SliceStable(values, func(i, j int) bool { return less(values[i].RelativePath, values[j].RelativePath) })
		conflict := PresetConflict{ModKey: key, Candidates: make([]PresetConflictCandidate, len(values))}
		for i, item := range values {
			conflict.Candidates[i] = PresetConflictCandidate{
				ActualPath: item.ActualPath, RelativePath: item.RelativePath,
				FolderName: item.FolderName, IsEnabled: item.IsEnabled,
			}
		}
		result = append(result, conflict)
	}
	less := newLocaleLess()
	sort.SliceStable(result, func(i, j int) bool { return less(result[i].ModKey, result[j].ModKey) })
	return result
}

func resolveSnapshotConflicts(conflicts []PresetConflict) error {
	for _, conflict := range conflicts {
		enabled := 0
		for _, candidate := range conflict.Candidates {
			if candidate.IsEnabled {
				enabled++
			}
		}
		disabledSeen := 0
		for _, candidate := range conflict.Candidates {
			if candidate.IsEnabled {
				continue
			}
			disabledSeen++
			if enabled == 0 && disabledSeen == 1 {
				continue
			}
			baseName := restoreDisabledPrefix(filepath.Base(candidate.ActualPath), candidate.FolderName)
			if _, err := renameUnique(candidate.ActualPath, baseName); err != nil {
				return err
			}
		}
	}
	return nil
}

func presetFromRow(row db.ModPresetRow) Preset {
	return Preset{
		ID: row.ID, Game: row.Game, Name: row.Name, Description: row.Description,
		CreatedAt: row.CreatedAt, UpdatedAt: row.UpdatedAt, Version: row.Version,
		IsLegacy: row.Version < modPresetVersion,
	}
}

func newPresetID() (string, error) {
	buffer := make([]byte, 16)
	if _, err := rand.Read(buffer); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buffer), nil
}
